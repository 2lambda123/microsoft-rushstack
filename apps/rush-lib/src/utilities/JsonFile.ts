// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as fsx from 'fs-extra';
import * as os from 'os';
import * as jju from 'jju';
import Utilities from './Utilities';

/**
 * Options for JsonFile.saveJsonFile()
 *
 * @public
 */
export interface ISaveJsonFileOptions {
  /**
   * If there is an existing file, and the contents have not changed, then
   * don't write anything; this preserves the old timestamp.
   */
  onlyIfChanged?: boolean;
}

/**
 * Utilities for reading/writing JSON files.
 * @public
 */
export default class JsonFile {
  /**
   * Loads a JSON file.
   */
  // tslint:disable-next-line:no-any
  public static loadJsonFile(jsonFilename: string): any {
    if (!fsx.existsSync(jsonFilename)) {
      throw new Error(`Input file not found: ${jsonFilename}`);
    }

    const buffer: Buffer = fsx.readFileSync(jsonFilename);
    try {
      const jsonData: Object = jju.parse(buffer.toString());
      return jsonData;
    } catch (error) {
      throw new Error(`Error reading "${jsonFilename}":` + os.EOL + `  ${error.message}`);
    }
  }

  // tslint:disable-next-line:no-any
  public static normalize(jsonData: any): string {
    const stringified: string = JSON.stringify(jsonData, undefined, 2) + '\n';
    return Utilities.getAllReplaced(stringified, '\n', '\r\n');
  }

  /**
   * Saves the file to disk.  Returns false if nothing was written due to options.onlyIfChanged.
   */
  // tslint:disable-next-line:no-any
  public static saveJsonFile(jsonData: any, jsonFilename: string, options: ISaveJsonFileOptions = {}): boolean {
    const normalized: string = JsonFile.normalize(jsonData);
    const buffer: Buffer = new Buffer(normalized); // utf8 encoding happens here

    if (options.onlyIfChanged) {
      // Has the file changed?
      if (fsx.existsSync(jsonFilename)) {
        try {
          const oldBuffer: Buffer = fsx.readFileSync(jsonFilename);
          if (Buffer.compare(buffer, oldBuffer) === 0) {
            // Nothing has changed, so don't touch the file
            return false;
          }
        } catch (error) {
          // Ignore this error, and try writing a new file.  If that fails, then we should report that
          // error instead.
        }
      }
    }

    fsx.writeFileSync(jsonFilename, buffer);

    // TEST CODE: Used to verify that onlyIfChanged isn't broken by a hidden transformation during saving.
    /*
    const oldBuffer2: Buffer = fsx.readFileSync(jsonFilename);
    if (Buffer.compare(buffer, oldBuffer2) !== 0) {
      console.log('new:' + buffer.toString('hex'));
      console.log('old:' + oldBuffer2.toString('hex'));

      throw new Error('onlyIfChanged logic is broken');
    }
    */

    return true;
  }
}
