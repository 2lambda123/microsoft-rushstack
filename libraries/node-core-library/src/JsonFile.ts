// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as fsx from 'fs-extra';
import * as os from 'os';
import * as jju from 'jju';

import { JsonSchema, IJsonSchemaErrorInfo, IJsonSchemaValidateOptions } from './JsonSchema';

/**
 * Options for JsonFile.stringify()
 *
 * @public
 */
export interface IJsonFileStringifyOptions {
  /**
   * If true, then "\n" will be used for newlines instead of the default "\r\n".
   */
  unixNewlines?: boolean;
}

/**
 * Options for JsonFile.saveJsonFile()
 *
 * @public
 */
export interface IJsonFileSaveOptions extends IJsonFileStringifyOptions {
  /**
   * If there is an existing file, and the contents have not changed, then
   * don't write anything; this preserves the old timestamp.
   */
  onlyIfChanged?: boolean;
}

/**
 * Utilities for reading/writing JSON files.
 * @public
 */
export class JsonFile {
  /**
   * Loads a JSON file.
   */
  public static load(jsonFilename: string): any { // tslint:disable-line:no-any
    if (!fsx.existsSync(jsonFilename)) {
      throw new Error(`Input file not found: ${jsonFilename}`);
    }

    const buffer: Buffer = fsx.readFileSync(jsonFilename);
    try {
      return jju.parse(buffer.toString());
    } catch (error) {
      throw new Error(`Error reading "${jsonFilename}":` + os.EOL + `  ${error.message}`);
    }
  }

  /**
   * Loads a JSON file and validate its schema.
   */
  public static loadAndValidate(jsonFilename: string, jsonSchema: JsonSchema,
    options?: IJsonSchemaValidateOptions): any { // tslint:disable-line:no-any

    const jsonObject: any = JsonFile.load(jsonFilename); // tslint:disable-line:no-any
    jsonSchema.validateObject(jsonObject, jsonFilename, options);

    return jsonObject;
  }

  /**
   * Loads a JSON file and validate its schema, reporting errors using a callback
   * @remarks
   * See JsonSchema.validateObjectWithCallback() for more info.
   */
  public static loadAndValidateWithCallback(jsonFilename: string, jsonSchema: JsonSchema,
    errorCallback: (errorInfo: IJsonSchemaErrorInfo) => void): any { // tslint:disable-line:no-any

    const jsonObject: any = JsonFile.load(jsonFilename); // tslint:disable-line:no-any
    jsonSchema.validateObjectWithCallback(jsonObject, errorCallback);

    return jsonObject;
  }

  /**
   * Serializes the specified JSON object to a string buffer.
   * @param jsonObject - the object to be serialized
   * @param options - other settings that control serialization
   * @returns a JSON string, with newlines, and indented with two spaces
   */
  public static stringify(jsonObject: Object, options?: IJsonFileStringifyOptions): string {
    JsonFile.validateNoUndefinedMembers(jsonObject);
    const stringified: string = JSON.stringify(jsonObject, undefined, 2) + '\n';

    if (options && options.unixNewlines) {
      return stringified;
    } else {
      return JsonFile._getAllReplaced(stringified, '\n', '\r\n');
    }

  }

  /**
   * Saves the file to disk.  Returns false if nothing was written due to options.onlyIfChanged.
   * @param jsonObject - the object to be saved
   * @param jsonFilename - the file path to write
   * @param options - other settings that control how the file is saved
   * @returns false if ISaveJsonFileOptions.onlyIfChanged didn't save anything; true otherwise
   */
  public static save(jsonObject: Object, jsonFilename: string, options: IJsonFileSaveOptions = {}): boolean {
    const normalized: string = JsonFile.stringify(jsonObject, options);

    const buffer: Buffer = new Buffer(normalized); // utf8 encoding happens here

    if (options.onlyIfChanged) {
      // Has the file changed?
      if (fsx.existsSync(jsonFilename)) {
        try {
          const oldBuffer: Buffer = fsx.readFileSync(jsonFilename);
          if (Buffer.compare(buffer, oldBuffer) === 0) {
            // Nothing has changed, so don't touch the file
            return false;
          }
        } catch (error) {
          // Ignore this error, and try writing a new file.  If that fails, then we should report that
          // error instead.
        }
      }
    }

    fsx.writeFileSync(jsonFilename, buffer);

    // TEST CODE: Used to verify that onlyIfChanged isn't broken by a hidden transformation during saving.
    /*
    const oldBuffer2: Buffer = fsx.readFileSync(jsonFilename);
    if (Buffer.compare(buffer, oldBuffer2) !== 0) {
      console.log('new:' + buffer.toString('hex'));
      console.log('old:' + oldBuffer2.toString('hex'));

      throw new Error('onlyIfChanged logic is broken');
    }
    */
    return true;
  }

  /**
   * Used to validate a data structure before writing.  Reports an error if there
   * are any undefined members.
   */
  // tslint:disable-next-line:no-any
  public static validateNoUndefinedMembers(jsonObject: Object): void {
    if (!jsonObject) {
      return;
    }
    if (typeof jsonObject === 'object') {
      for (const key of Object.keys(jsonObject)) {
        // tslint:disable-next-line:no-any
        const value: any = jsonObject[key];
        if (value === undefined) {
          throw new Error(`The key "${key}" is undefined`);
        }
        JsonFile.validateNoUndefinedMembers(value);
      }
    }
  }

  /**
   * Returns the same thing as targetString.replace(searchValue, replaceValue), except that
   * all matches are replaced, rather than just the first match.
   * @param targetString  The string to be modified
   * @param searchValue   The value to search for
   * @param replaceValue  The replacement text
   */
  private static _getAllReplaced(targetString: string, searchValue: string, replaceValue: string): string {
    return targetString.split(searchValue).join(replaceValue);
  }
}
