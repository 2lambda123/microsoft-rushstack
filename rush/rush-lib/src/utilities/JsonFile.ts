// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as fsx from 'fs-extra';
import * as os from 'os';
import * as jju from 'jju';
import Utilities from './Utilities';

/**
 * Utilities for reading/writing JSON files.
 * @public
 */
export default class JsonFile {

  /* tslint:disable:no-any */ // JSON objects are dynamically typed
  public static loadJsonFile(jsonFilename: string): any {
    if (!fsx.existsSync(jsonFilename)) {
      throw new Error(`Input file not found: ${jsonFilename}`);
    }

    const buffer: Buffer = fsx.readFileSync(jsonFilename);
    try {
      return jju.parse(buffer.toString());
    } catch (error) {
      throw new Error(`Error reading "${jsonFilename}":` + os.EOL + `  ${error.message}`);
    }
  }

  public static saveJsonFile(jsonData: any, jsonFilename: string): void {
    const stringified: string = JSON.stringify(jsonData, undefined, 2) + '\n';
    const normalized: string = Utilities.getAllReplaced(stringified, '\n', '\r\n');
    fsx.writeFileSync(jsonFilename, normalized);
  }
  /* tslint:enable:no-any */
}
