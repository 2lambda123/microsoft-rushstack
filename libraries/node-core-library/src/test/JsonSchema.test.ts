// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { JsonFile, type JsonObject } from '../JsonFile';
import { JsonSchema, type IJsonSchemaErrorInfo } from '../JsonSchema';

describe(JsonSchema.name, () => {
  const schemaPath: string = `${__dirname}/test-data/test-schema.json`;
  const schema: JsonSchema = JsonSchema.fromFile(schemaPath);

  describe('loadAndValidate', () => {
    test('successfully validates a JSON file', () => {
      const jsonPath: string = `${__dirname}/test-data/test.json`;
      const jsonObject: JsonObject = JsonFile.loadAndValidate(jsonPath, schema);

      expect(jsonObject).toMatchObject({
        exampleString: 'This is a string',
        exampleArray: ['apple', 'banana', 'coconut']
      });
    });

    test('successfully validates a JSON file against a draft-04 schema', () => {
      const schemaPathDraft04: string = `${__dirname}/test-data/test-schema-draft-04.json`;
      const schemaDraft04: JsonSchema = JsonSchema.fromFile(schemaPathDraft04);

      const jsonPath: string = `${__dirname}/test-data/test.json`;
      const jsonObject: JsonObject = JsonFile.loadAndValidate(jsonPath, schemaDraft04);

      expect(jsonObject).toMatchObject({
        exampleString: 'This is a string',
        exampleArray: ['apple', 'banana', 'coconut']
      });
    });

    test('validates a JSON file against a draft-07 schema', () => {
      const schemaPathDraft07: string = `${__dirname}/test-data/test-schema-draft-07.json`;
      const schemaDraft07: JsonSchema = JsonSchema.fromFile(schemaPathDraft07);

      const jsonPath: string = `${__dirname}/test-data/test.json`;
      const jsonObject: JsonObject = JsonFile.loadAndValidate(jsonPath, schemaDraft07);

      expect(jsonObject).toMatchObject({
        exampleString: 'This is a string',
        exampleArray: ['apple', 'banana', 'coconut']
      });
    });

    test('validates a JSON file using nested schemas', () => {
      const schemaPathChild: string = `${__dirname}/test-data/test-schema-nested-child.json`;
      const schemaChild: JsonSchema = JsonSchema.fromFile(schemaPathChild);

      const schemaPathNested: string = `${__dirname}/test-data/test-schema-nested.json`;
      const schemaNested: JsonSchema = JsonSchema.fromFile(schemaPathNested, {
        dependentSchemas: [schemaChild]
      });

      const jsonPath: string = `${__dirname}/test-data/test.json`;
      const jsonObject: JsonObject = JsonFile.loadAndValidate(jsonPath, schemaNested);

      expect(jsonObject).toMatchObject({
        exampleString: 'This is a string',
        exampleArray: ['apple', 'banana', 'coconut']
      });
    });
  });

  describe('validateObjectWithCallback', () => {
    test('successfully reports a compound validation error', () => {
      const jsonPath2: string = `${__dirname}/test-data/test2.json`;
      const jsonObject2: JsonObject = JsonFile.load(jsonPath2);

      const errorDetails: string[] = [];
      schema.validateObjectWithCallback(jsonObject2, (errorInfo: IJsonSchemaErrorInfo) => {
        errorDetails.push(errorInfo.details);
      });

      expect(errorDetails).toMatchSnapshot();
    });
  });
});
