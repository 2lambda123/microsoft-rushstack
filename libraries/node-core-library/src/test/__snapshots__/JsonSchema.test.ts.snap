// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonSchema validateObjectWithCallback successfully reports a compound validation error 1`] = `
Array [
  "
Error: #/definitions/type1/required (must have required property 'field1')
       must have required property 'field1'
Error: #/definitions/type2/required (must have required property 'field3')
       must have required property 'field3'
Error: #/properties/exampleOneOf/oneOf (must match exactly one schema in oneOf)
       must match exactly one schema in oneOf",
]
`;
